### PR TITLE
feat(notifications): add ability to clear notifications

### PR DESCRIPTION
**Pull request by gep13**
- Upon entering a room, clear all the notifications for that room, as
the messages have now been read

### CODE REVIEW

**Comment by gep13** [comment]
@Odonno what are your thoughts on this?

Ideally, I would like to see all the notifications, i.e. in the Action Center disappear, once I have entered the room, instead of having to dismiss them manually.

When using the ToastNotification within a UAP application, there doesn't seem to be the option to use the Tag and Group properties, but by creating a concrete implementation of the ILocalNotificationService specifically for Windows Phone, these options become available.

All that is left in this PR to complete the implementation would be to call ClearNotifications from the Page Loaded event.  I have done this in the past in a Windows Phone Silverlight Application using:

```
    <i:Interaction.Triggers>
        <i:EventTrigger EventName="Loaded">
            <i:InvokeCommandAction Command="{Binding PageLoadedCommand}" />
        </i:EventTrigger>
    </i:Interaction.Triggers>
```

and then:

```
        public RelayCommand PageLoadedCommand
        {
            get
            {
                return new RelayCommand(this.PageLoaded);
            }
        }
```

Where in PageLoaded you would have the call to ClearNotifications, however, this relies on pulling in:

`xmlns:i="clr-namespace:System.Windows.Interactivity;assembly=System.Windows.Interactivity"`

which I believe it the "old" way to do it.

If you like this idea of removing the notifications from the Action Center, can you suggest a good way to make the call to ClearNotifications?

Thanks!

**Comment by Odonno** [comment]
Yeah ! Great !

It was a time I searched for a way to remove Toast Notifications on both platforms, using UAP. If you find it works only on WP8.1, fine, we'll take a look at that.

About the use of the `PageLoaded` event, I think it's a bad idea. I presume you want to remove notifications when reading a room ? So, I have the dedicated command inside MainViewModel (`SelectRoomCommand`).

File: Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+{
+    public class WindowsPhoneNotificationService : ILocalNotificationService
+    {
+        public void SendNotification(string title, string content, string id = null)
+        {
+            // Send any notification (Tile, Toast)
+            SendToastNotification(title, content, id);
+        }
+
+        public async void ClearNotificationGroup(string id)
```
Comment (Odonno) [inline Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs]: Should be `async Task`...

**Comment by Odonno** [comment]
First, it could be great to rewrite your implementation of ToastNotificationService to follow the DRY principle. Maybe an abstract base class on the `SHARED` and specific code in frontend projects ?

File: Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs (hunk 1 above)
Comment (gep13) [inline Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs]: Ah, good catch :+1:

**Comment by gep13** [comment]
> First, it could be great to rewrite your implementation of ToastNotificationService to follow the DRY principle. Maybe an abstract base class on the SHARED and specific code in frontend projects ?

Yes, that makes a lot of sense, as there is a lot of repetition.   Now that I know you are not against the idea, I will spend more time fixing up the PR.

Thanks!

**Comment by gep13** [comment]
>So, I have the dedicated command inside MainViewModel (SelectRoomCommand).

This sounds perfect, thanks!

File: Gitter/Gitter.Configuration/Notifications.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Gitter.Configuration
+{
+    public static class Notifications
```
Comment (Odonno) [inline Gitter/Gitter.Configuration/Notifications.cs]: Should be `NotificationConstants`

File: Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Gitter.Services.Concrete
+{
+    public class WindowsPhoneNotificationService : BaseNotificationService, ILocalNotificationService
+    {
+        public override void SendNotification(string title, string content, string id = null, string group = null)
+        {
+            var notification = this.CreateToastNotification(title, content, id);
+
+#if WINDOWS_PHONE_APP
```
Comment (Odonno) [inline Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs]: No need of `#if` anymore

File: Gitter/Gitter.Configuration/Notifications.cs (hunk 2 above)
Comment (gep13) [inline Gitter/Gitter.Configuration/Notifications.cs]: Ok, that makes more sense.  Done.

File: Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs (hunk 3 above)
Comment (gep13) [inline Gitter/Gitter/Gitter.WindowsPhone/Services/WindowsPhoneNotificationService.cs]: This was me attempting to get it to work under Windows Phone, but even if the compiler condition in place, it still wasn't finding the Tag and Group properties.  I have removed them for now.

**Comment by gep13** [comment]
@Odonno how are you? I was just wondering if you had any thoughts on the best approach to get this working?  It currently doesn't compile for me.  Does it work for you?

Thanks!

**Comment by Odonno** [comment]
@gep13 Sorry, I do not have time currently. I know what to do but it's a bit difficult to describe it.

File: Gitter/Gitter.Services/Concrete/LocalNotificationService.cs [hunk 4]
```diff
@@ -1,54 +1,20 @@
...
     {
-        public void SendNotification(string title, string content, string id = null)
+        public override void SendNotification(string title, string content, string id = null, string group = null)
         {
-            // Send any notification (Tile, Toast)
-            SendToastNotification(title, content, id);
-        }
+            var notification = this.CreateToastNotification(title, content, id);
 
+            this.toastNotifier.Show(notification);
```
Comment (Odonno) [inline Gitter/Gitter.Services/Concrete/LocalNotificationService.cs]: Remove `this` whenever you can, here you can

**Comment by gep13** [comment]
>@gep13 Sorry, I do not have time currently. I know what to do but it's a bit difficult to describe it.

Not a problem at all, just wanted to check :+1:

**Comment by Odonno** [comment]
I have done the job. I think it's done, what do you think ?

**Comment by gep13** [comment]
@Odonno does the project currently compile for you?  When I try, I am seeing three compilation errors:

**Comment by gep13** [comment]
@Odonno ah, hold on...

Looks like you created a new branch, rather than pushing your changes into the existing one:

image

Let me have another play...

**Comment by Odonno** [comment]
@gep13 Oh sorry, yes, I mis-using some git commands and I have created a branch gep13/ClearNotifications. Please consider remove it, it is a useless branch.

**Comment by Odonno** [comment]
I do know why you have these errors... Stil, I removed the class LocalNotificationService. Are you sure you have the last version of the PR ?

**Comment by Odonno** [comment]
In *NotificationsBackgroundTask*, we can't use *LocalNotificationService* anymore since I removed it to make specific service. And we can't use the new services anymore as the *Tasks* project can't target the frontend...

**Comment by Odonno** [comment]
I have done something on another project a long time ago. We need to make a lot of projects to achieve our goal : 

* create two projects for Tasks : both Windows and Windows Phone
* create two project for Services : both Windows and Windows Phone
* and then reference projects in a way we can call services from tasks

What do you think ?

**Comment by gep13** [comment]
> Please consider remove it, it is a useless branch.

Ok, will do.

**Comment by gep13** [comment]
>I have done something on another project a long time ago. We need to make a lot of projects to achieve our goal : 
•create two projects for Tasks : both Windows and Windows Phone
•create two project for Services : both Windows and Windows Phone
•and then reference projects in a way we can call services from tasks
What do you think ?

I agree this sounds like a lot of "extra" work, but I also think it is necessary.  Would you like me to start on this, or is this something that you want to take care of?

**Comment by Odonno** [comment]
Nice !

I let you do the job. 
If you have questions, I am here ! 
In case you need help on the code, I'll check that on saturday certainly.

**Comment by Odonno** [comment]
I've done some refactoring on the code (between NuGet package and the project).
So, I've created both projects for Tasks (Windows & Windows Phone).

Can we merge the code now and I'll do the last things to make it work ?

**Comment by gep13** [comment]
Yip, I have no issues with that. Thanks for looking into this!  You want me to rebase on to the latest version of your branch?  Currently it is saying there are conflicts, or are you ok to look into this?

**Comment by Odonno** [comment]
I'm on it !

**Comment by gep13** [comment]
:+1:

**Comment by Odonno** [comment]
Oh ! I mispelled your branch name and so I merged the code from another branch that contains the entire code...don't why it worked but it is a big mistake.

I close this PR but it is a workful PR, not a falied/abandonned PR.

Thank you @gep13  !

**Pull request closed without merging**